### PR TITLE
add benchmarks for different serialization frameworks

### PR DESCRIPTION
**Pull request by NicoK**
FYI: These are the results with Flink 1.6.0 on my machine:
```
Benchmark                                                       Mode  Cnt    Score    Error   Units
SerializationFrameworkBenchmarks.testAvroForcedSerializer      thrpt   30  115.814 ±  2.714  ops/ms
SerializationFrameworkBenchmarks.testAvroSchemaSerializer      thrpt   30  160.859 ±  4.622  ops/ms
SerializationFrameworkBenchmarks.testKryoSerializer            thrpt   30   94.518 ±  1.939  ops/ms
SerializationFrameworkBenchmarks.testKryoSerializerRegistered  thrpt   30  120.472 ± 12.675  ops/ms
SerializationFrameworkBenchmarks.testPojoSerializer            thrpt   30  234.788 ±  5.263  ops/ms
SerializationFrameworkBenchmarks.testPojoSerializerRegistered  thrpt   30  256.074 ± 16.486  ops/ms
SerializationFrameworkBenchmarks.testRowSerializer             thrpt   30  297.823 ± 20.833  ops/ms
SerializationFrameworkBenchmarks.testTupleSerializer           thrpt   30  313.929 ±  9.551  ops/ms
```

### CODE REVIEW

File: src/main/java/org/apache/flink/benchmark/functions/BaseSourceWithKeyRange.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+	@Override
+	public void run(SourceContext<T> out) {
+		int keyId = 0;
+		while (--remainingEvents >= 0) {
+			T element = getElement(keyId);
+			synchronized (out.getCheckpointLock()) {
+				out.collect(element);
+			}
+			++keyId;
```
Comment (pnowojski) [inline src/main/java/org/apache/flink/benchmark/functions/BaseSourceWithKeyRange.java:53]: I like the spirit of extracting every single instruction out of the lock, but maybe this `getElement ` and `++key` are a bit too much ;)

File: src/main/java/org/apache/flink/benchmark/SerializationFrameworkBenchmarks.java [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+import org.openjdk.jmh.runner.options.VerboseMode;
+
+import java.util.Arrays;
+
+/**
+ * Benchmark for serializing POJOs and Tuples with different serialization frameworks.
+ */
+public class SerializationFrameworkBenchmarks extends BenchmarkBase {
+
+	private static final int RECORDS_PER_INVOCATION = 100_000;
```
Comment (pnowojski) [inline src/main/java/org/apache/flink/benchmark/SerializationFrameworkBenchmarks.java]: `100_000` records per invocation and judging from the results that you posted one invocation takes ~1 second, right? If you double the value, the results stay the same?

File: src/main/java/org/apache/flink/benchmark/SerializationFrameworkBenchmarks.java [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+		env.addSource(new PojoSource(RECORDS_PER_INVOCATION, 10))
+				.rebalance()
+				.addSink(new DiscardingSink<>());
+
+		env.execute();
+	}
+
+	@Benchmark
+	@OperationsPerInvocation(value = SerializationFrameworkBenchmarks.RECORDS_PER_INVOCATION)
+	public void testPojoSerializerRegistered() throws Exception {
```
Comment (pnowojski) [inline src/main/java/org/apache/flink/benchmark/SerializationFrameworkBenchmarks.java]: why registering is faster?

File: src/main/java/org/apache/flink/benchmark/SerializationFrameworkBenchmarks.java [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+		env.addSource(new PojoSource(RECORDS_PER_INVOCATION, 10))
+				.rebalance()
+				.addSink(new DiscardingSink<>());
+
+		env.execute();
+	}
+
+	@Benchmark
+	@OperationsPerInvocation(value = SerializationFrameworkBenchmarks.RECORDS_PER_INVOCATION)
+	public void testAvroSchemaSerializer() throws Exception {
```
Comment (pnowojski) [inline src/main/java/org/apache/flink/benchmark/SerializationFrameworkBenchmarks.java]: please drop the `test` prefix from the methods. It's redundant with `@Benchmark` annotation and longer names are more problematic for `codespeed` ui.

**Review by pnowojski — COMMENTED** [review]
More or less the change LGTM.

I have two concerns:
- records that you are serialising are array "heavy", thus arrays overheads could be dominant factor when comparing avro, pojo, row, tuple and kryo
- do we need `testAvroForcedSerializer`, `testKryoSerializer` and `testPojoSerializer`? How likely is that we will brake something for them, while not braking for registered kryo/pojo? (Asking because I would like to limit number of benchmarks)

File: src/main/java/org/apache/flink/benchmark/functions/BaseSourceWithKeyRange.java (hunk 1 above)
Comment (NicoK) [reply src/main/java/org/apache/flink/benchmark/functions/BaseSourceWithKeyRange.java:53]: I don't see a problem here...both accesses are only from a single thread anyway, so no locking required

File: src/main/java/org/apache/flink/benchmark/SerializationFrameworkBenchmarks.java (hunk 3 above)
Comment (NicoK) [reply src/main/java/org/apache/flink/benchmark/SerializationFrameworkBenchmarks.java]: I'm assuming, then we also do not use full class names in the serialization format but rather registered tags

File: src/main/java/org/apache/flink/benchmark/SerializationFrameworkBenchmarks.java (hunk 2 above)
Comment (NicoK) [reply src/main/java/org/apache/flink/benchmark/SerializationFrameworkBenchmarks.java]: true - they don't :(
with `300_000` I get the following results:
```
SerializationFrameworkBenchmarks.testAvroForcedSerializer      thrpt   30  129.923 ±  3.274  ops/ms
SerializationFrameworkBenchmarks.testAvroSchemaSerializer      thrpt   30  216.473 ± 34.747  ops/ms
SerializationFrameworkBenchmarks.testKryoSerializer            thrpt   30   70.413 ±  1.143  ops/ms
SerializationFrameworkBenchmarks.testKryoSerializerRegistered  thrpt   30  104.970 ±  1.541  ops/ms
SerializationFrameworkBenchmarks.testPojoSerializer            thrpt   30  344.357 ± 25.534  ops/ms
SerializationFrameworkBenchmarks.testPojoSerializerRegistered  thrpt   30  344.299 ± 22.487  ops/ms
SerializationFrameworkBenchmarks.testRowSerializer             thrpt   30  416.346 ± 29.662  ops/ms
SerializationFrameworkBenchmarks.testTupleSerializer           thrpt   30  440.312 ±  5.931  ops/ms
```
it's a strange thing that Kryo actually got slower (as the only benchmark!)

Comment (NicoK) [reply src/main/java/org/apache/flink/benchmark/SerializationFrameworkBenchmarks.java]: looks like 300k is a good enough number: 100k and 200k have significantly lower numbers for Pojo, Row, and Tuple while the gain by 400k and 500k is not significant any more (at least on my PC)

**Comment by NicoK** [comment]
done - pushed the changes as agreed

**Review by pnowojski — APPROVED** [review]
LGTM

**Pull request merged**